fedimint-server/src/config/api.rs

```diff
@@ -37,27 +47,157 @@ impl ConfigApi {
     }
 
     // Sets the auth and decryption key derived from the password
-    pub fn set_password(&self, auth: ApiAuth) -> Result<(), ApiError> {
+    pub fn set_password(&self, auth: ApiAuth) -> Result<()> {
         let mut state = self.state.lock().expect("lock poisoned");
 
-        if *state != ConfigApiState::SetPassword {
-            Err(ApiError::bad_request("Password already set".to_string()))
-        } else {
-            *state = ConfigApiState::SetPeerConnectionInfo(auth);
-            Ok(())
+        match &*state {
+            ConfigApiState::SetPassword => *state = ConfigApiState::SetConnections(auth),
+            _ => return Err(ApiError::bad_request("Password already set".to_string())),
+        }
+
+        Ok(())
+    }
+
+    /// Sets our connection info, possibly sending it to a leader
+    pub async fn set_config_gen_connections(
+        &self,
+        request: ConfigGenConnectionsRequest,
+    ) -> Result<()> {
+        let connection = {
+            let mut state = self.state.lock().expect("lock poisoned");
+
+            let connection = match (*state).clone() {
+                ConfigApiState::SetConnections(auth) => ConfigGenConnections::new(request, auth)?,
+                ConfigApiState::SetConfigGenParams(old) => old.with_request(request),
+                _ => return Err(ApiError::bad_request("Config already set".to_string())),
+            };
+            *state = ConfigApiState::SetConfigGenParams(connection.clone());
+
+            connection
+        };
+
+        if let Some(url) = connection.request.leader_api_url.clone() {
+            // Note PeerIds don't really exist at this point, but id doesn't matter because
+            // it's not used in the WS client for anything, perhaps it should be removed
+            let client = WsAuthenticatedClient::new(url, PeerId::from(0), connection.auth.clone());
+            client
+                .add_config_gen_peer(connection.as_peer_info())
+                .await
+                .map_err(|_| ApiError::not_found("Unable to connect to the leader".to_string()))?;
+        }
+
+        Ok(())
+    }
+
+    /// Called from `set_config_gen_connections` to add a peer's connection info
+    /// to the leader
+    pub fn add_config_gen_peer(&self, peer: PeerServerParams) -> Result<()> {
+        let mut state = self.state.lock().expect("lock poisoned");
+
+        let mut connection = match &*state {
+            ConfigApiState::SetConfigGenParams(connection) => connection.clone(),
+            _ => return Self::bad_request("Not ready to receive peer server params"),
+        };
+        connection.peers.insert(peer.cert.clone(), peer);
+        *state = ConfigApiState::SetConfigGenParams(connection);
+
+        Ok(())
+    }
+
+    /// Returns the peers that have called `add_config_gen_peer` on the leader,
+    /// only the leader will need to call this
```

Comment:
Should this be a subscription based API call? Like, give me all `PeerServerParams` every time they change? That way everyone sees the others connect in real time. There should be another endpoint possibly that returns the full params once the leader clicks a buttons indicating everyone who was expected connected.

Reply:
Yes there will be an API endpoint to return all the configs.  This endpoint I thought would only be used by the leader for displaying in their UI.  The followers would only see that the leader is online and accepted their request, then they would wait for the leader to configure the rest of the config gen params.

What is a good example of the subscription based API call?

Reply:
There are a few endpoints @Maan2003 refactored to wait for a certain event to occur, although none of these is subscription-based yet afaik. E.g. waiting for a transaction to be confirmed or rejected:

https://github.com/fedimint/fedimint/blob/cf9917ef806e19890d0073ddb8e9efe7b6c48b65/fedimint-server/src/net/api.rs#L228-L239

You couldn't rely on the DB and would probably need to use [`Notify`](https://dtantsur.github.io/rust-openstack/tokio/sync/struct.Notify.html) directly.

Reply:
If we let peers call this and want it to block I'm not sure the best way to implement...I guess we'd need a `Map<TlsCert, Notify>` and require peers to pass in their cert?  I'm tempted to leave this for now until we know how the UI will be calling it.

Reply:
I think one notify would be good enough? I was assuming we want to give at least the leader a live view of who is connected, so instead of polling this fn would return each time someone connects.

One possible design:
1. client sends how many peers they are aware of
2. server subscribes to notify
3. server checks if current number of peers is more than what client said, return peers to client if so
4. server waits for notification, then jump to 3